datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,114 @@
+      Optional<Integer> partitionsPerTask, Optional<Integer> partitionFullnessFactorPct, Optional<ZkClient> zkClient,
+      String clusterName) {
+    super(maxTasks, imbalanceThreshold, maxPartitionPerTask, enableElasticTaskAssignment, partitionsPerTask,
+        partitionFullnessFactorPct, zkClient, clusterName);
+    _throughputProvider = throughputProvider;
+    _sourceClusterResolver = sourceClusterResolver;
+  }
+
+  @Override
+  public Map<String, Set<DatastreamTask>> assignPartitions(Map<String, Set<DatastreamTask>> currentAssignment,
+      DatastreamGroupPartitionsMetadata datastreamPartitions) {
+    DatastreamGroup datastreamGroup = datastreamPartitions.getDatastreamGroup();
+    String datastreamGroupName = datastreamGroup.getName();
+    List<String> assignedPartitions = getAssignedPartitionsForDatastreamGroup(currentAssignment, datastreamGroupName);
+
+    // Do throughput based assignment only initially, when no partitions have been assigned yet
+    if (!assignedPartitions.isEmpty()) {
+      return super.assignPartitions(currentAssignment, datastreamPartitions);
+    }
+
+    Map<String, ClusterThroughputInfo> partitionThroughputInfo;
+    // Attempting to retrieve partition throughput info with a fallback mechanism to StickyPartitionAssignmentStrategy
+    try {
+      partitionThroughputInfo = fetchPartitionThroughputInfo();
+    } catch (RetriesExhaustedException ex) {
+      LOG.warn("Attempts to fetch partition throughput timed out. Falling back to regular partition assignment strategy");
+      return super.assignPartitions(currentAssignment, datastreamPartitions);
+    }
+
+    LOG.info("Old partition assignment info, assignment: {}", currentAssignment);
+    Validate.isTrue(currentAssignment.size() > 0,
+        "Zero tasks assigned. Retry leader partition assignment.");
+
+    // TODO Get task count estimate and perform elastic task count validation
+    // TODO Get task count estimate based on throughput and pick a winner
+    int maxTaskCount = 0;
+
+    // TODO Get unassigned partitions
+    // Calculating unassigned partitions
+    List<String> unassignedPartitions = new ArrayList<>();
+
+    // Resolving cluster name from datastream group
+    String clusterName = _sourceClusterResolver.getSourceCluster(datastreamPartitions.getDatastreamGroup());
+    ClusterThroughputInfo clusterThroughputInfo = partitionThroughputInfo.get(clusterName);
+
+    // Doing assignment
+    LoadBasedPartitionAssigner partitionAssigner = new LoadBasedPartitionAssigner();
+    return partitionAssigner.assignPartitions(clusterThroughputInfo, assignedPartitions, unassignedPartitions,
+        maxTaskCount);
+  }
+
+  private Map<String, ClusterThroughputInfo> fetchPartitionThroughputInfo() {
+    return PollUtils.poll(() -> {
+      try {
+        return _throughputProvider.getThroughputInfo();
+      } catch (Exception ex) {
+        LOG.warn("Failed to fetch partition throughput info.");
```

Comment:
nit: may be interested in exception. Also retry count in log will be helpful. 

You can pick this in next PR.

Reply:
Added a todo comment